delta_backend/src/ConversionChecker.py

```diff
@@ -85,25 +93,120 @@ def convertData(self, expressionType, expressionRule, fieldName, fieldValue):
                     expressionRule, fieldName, fieldValue, self.summarise, self.report_unexpected_exception
                 )
             case _:
-                return "Schema expression not found! Check your expression type : " + expressionType
+                raise ValueError("Schema expression not found! Check your expression type : " + expressionType) 
+
+    # Utility function for logging errors
+    def _log_error(self, fieldName, fieldValue, e, code=ExceptionMessages.RECORD_CHECK_FAILED):
+        if isinstance(e, Exception):
+            message = ExceptionMessages.MESSAGES[ExceptionMessages.UNEXPECTED_EXCEPTION] % (e.__class__.__name__, str(e))
+        else:
+            message = str(e)  # if a simple string message was passed 
+
+        # Ensure message-level deduplication
+        if any(existing.get("message") == message for existing in self.errorRecords):
+            return 
+
+        self.errorRecords.append({
+            "code": code,
+            "field": fieldName,
+            "value": fieldValue,
+            "message": message
+        })
 
-    # Convert ISO date string to a specific format (e.g. YYYYMMDD)
     def _convertToDate(self, expressionRule, fieldName, fieldValue, summarise, report_unexpected_exception):
         if not fieldValue:
             return ""
 
         if not isinstance(fieldValue, str):
+            if report_unexpected_exception:
+                self._log_error(fieldName, fieldValue, "Value is not a string")
             return ""
-        # Reject partial dates like "2024" or "2024-05"
-        if re.match(r"^\d{4}(-\d{2})?$", fieldValue):
+        
+        # Normalize expression rule
+        format_str = expressionRule.replace("format:", "").strip()
+        
+        # Reject partial ISO dates like "2024" or "2024-05"
+        if format_str == "%Y%m%d" and re.match(r"^\d{4}(-\d{2})?$", fieldValue):
+            if report_unexpected_exception:
+                self._log_error(fieldName, fieldValue, "Partial date not accepted")
             return ""
+
+        # Handle only the recorded field with extended ISO + timezone support
+        if fieldName == "recorded":
+            # Accept "YYYY-MM-DD" and return as is
+            if re.match(r"^\d{4}-\d{2}-\d{2}$", fieldValue):
+                try:
+                    dt = datetime.strptime(fieldValue, "%Y-%m-%d")
+                    if dt.date() > datetime.now(ZoneInfo("UTC")).date():
```

Comment:
Definitely an edge case, but what if we receive a vaccination event at say 00:30 BST? Then `now` in UTC would be 23:30 on the previous day and therefore the recorded date would appear to be in the future. (I think? 😅)